mesonbuild/ast/interpreter.py

```diff
@@ -492,11 +492,12 @@ def get_cur_value(self, var_name: str, allow_none: bool = False) -> T.Union[Base
             if len(self.nesting) >= len(nesting) and self.nesting[:len(nesting)] == nesting:
                 ret = value
                 break
-        if ret is None and allow_none:
-            return ret
-        if ret is None and self.tainted:
-            return UnknownValue()
-        assert ret is not None
+        if ret is None:
+            if self.tainted:
+                return UnknownValue()
+            if not allow_none:
+                path = mlog.get_relative_path(Path(self.current_node.filename), Path(os.getcwd()))
+                mlog.warning(f"{path}:{self.current_node.lineno}:{self.current_node.colno} will always crash if executed, since a variable named `{var_name}` is not defined")
```

Comment:
It doesn't seem like the test you have matches this? Wouldn't that need a test like:
```meson
if false
  message(variable_that_is_not_defined)
endif
```